File: code_samples/section9/lesson/section9.cs
#nullable enable

/*
    This file contains THREE classic dynamic-programming / recursion problems,
    plus a simple top-level test harness:

      1) MaxRootToLeaf (tree recursion / tree DP)
         - Returns the maximum sum along any path that starts at the root
           and ends at a LEAF node.
         - Uses recursion:
             best(node) = node.Val + max(best(node.Left), best(node.Right))

      2) TspBitmask (DP over subsets / bitmask DP)
         - Computes the minimum-cost PATH that:
             * starts at node 0
             * visits every node exactly once
             * ends at ANY node (does NOT return to 0)
         - DP state:
             dp[mask, j] = min cost to start at 0, visit nodes in 'mask',
                           and finish at node j.

      3) MatrixChainOrder (interval DP)
         - Given dimensions p = [p0, p1, ..., pn] for matrices A1..An
           where Ai has dimension p[i-1] x p[i],
           returns the minimum scalar multiplications to compute A1..An.

    Notes on C# / nullable:
      - #nullable enable turns on nullable reference type analysis.
      - TreeNode? means "TreeNode reference or null".
*/

/* ============================================================
   1) MaxRootToLeaf
   ============================================================ */

/*
    Returns the maximum root-to-leaf sum in a binary tree.

    Edge behavior:
      - If root is null, there is no valid path, so we return int.MinValue
        as a sentinel meaning "no path".
        This makes max() logic convenient: a missing subtree never wins
        against a real subtree sum.

    Base cases:
      - root == null                 => int.MinValue (no path)
      - root is a leaf (no children) => root.Val

    Recursive case:
      - compute best sum from left subtree
      - compute best sum from right subtree
      - pick the larger one and add root.Val

    Complexity:
      - Time:  O(number of nodes) because each node is visited once.
      - Space: O(height) due to recursion stack.
*/
static int MaxRootToLeaf(TreeNode? root) {
    // Empty tree => no root-to-leaf path exists.
    // Return sentinel minimal value so it won't be chosen by Math.Max
    // when a real path exists.
    if (root == null) {
        return int.MinValue;
    }

    // Leaf node => the only root-to-leaf path is the node itself.
    if (root.Left == null && root.Right == null) {
        return root.Val;
    }

    // Recursively compute best sums from each child.
    int leftBest = MaxRootToLeaf(root.Left);
    int rightBest = MaxRootToLeaf(root.Right);

    // Best path through this node = node value + best child path.
    return root.Val + System.Math.Max(leftBest, rightBest);
}

/* ============================================================
   2) TspBitmask (subset DP)
   ============================================================ */

/*
    Solves a TSP-like variant using bitmask DP.

    Input:
      - n: number of nodes (0..n-1)
      - dist: n x n matrix where dist[i,j] is the cost from i -> j

    Goal:
      - Start at node 0
      - Visit all nodes exactly once
      - End at ANY node (no "return to start" requirement)
      - Minimize total cost

    DP meaning:
      dp[mask, j] = minimum cost to:
        - start at node 0
        - visit exactly the nodes indicated by 'mask'
        - end at node j
      (j must be included in mask)

    Initialization:
      dp[1<<0, 0] = 0
      All other states start as INF (unreachable).

    Transition:
      From (mask, j), try to go to any unvisited node k:
        nextMask = mask | (1<<k)
        dp[nextMask, k] = min(dp[nextMask, k], dp[mask, j] + dist[j,k])

    Answer:
      fullMask = (1<<n) - 1
      min over j of dp[fullMask, j]

    Complexity:
      - Time:  O(n^2 * 2^n)
      - Space: O(n * 2^n)

    IMPORTANT:
      This returns the best *Hamiltonian path* starting at 0.
      If you wanted a *tour* (return to 0), you would compute:
        min_j dp[fullMask, j] + dist[j, 0]
*/
static int TspBitmask(int n, int[,] dist) {
    // Total number of subsets of nodes.
    int maxMask = 1 << n;

    // "Infinity" large value; divide by 4 to reduce overflow risk when adding.
    int INF = int.MaxValue / 4;

    // dp[mask, j] = best cost for subset 'mask' ending at node j.
    // We'll initialize every entry to INF (meaning "unreachable").
    var dp = new int[maxMask, n];
    for (int m = 0; m < maxMask; m++) {
        for (int j = 0; j < n; j++) {
            dp[m, j] = INF;
        }
    }

    // Base case: only node 0 visited, and we are currently at node 0.
    dp[1 << 0, 0] = 0;

    // Iterate through every subset of nodes.
    for (int mask = 0; mask < maxMask; mask++) {
        // Try every possible "current endpoint" j.
        for (int j = 0; j < n; j++) {

            // If node j is not included in mask, dp[mask,j] is invalid.
            if ((mask & (1 << j)) == 0) continue;

            int curr = dp[mask, j];

            // Skip unreachable states.
            if (curr == INF) continue;

            // Try to extend the path to some unvisited node k.
            for (int k = 0; k < n; k++) {
                // If k is already visited in mask, we can't visit it again.
                if ((mask & (1 << k)) != 0) continue;

                int nextMask = mask | (1 << k);
                int cand = curr + dist[j, k];

                // Standard DP "relaxation": keep the cheapest cost.
                if (cand < dp[nextMask, k]) {
                    dp[nextMask, k] = cand;
                }
            }
        }
    }

    // All nodes visited.
    int fullMask = maxMask - 1;

    // Best cost can end at any node j.
    int best = INF;
    for (int j = 0; j < n; j++) {
        if (dp[fullMask, j] < best) {
            best = dp[fullMask, j];
        }
    }
    return best;
}

/* ============================================================
   3) MatrixChainOrder (interval DP)
   ============================================================ */

/*
    Matrix chain multiplication optimization.

    Input:
      p: dimensions array of length n+1
         matrices are:
           A1: p[0] x p[1]
           A2: p[1] x p[2]
           ...
           An: p[n-1] x p[n]
      Here, n = p.Length - 1.

    DP state:
      dp[i, j] = minimum cost to multiply Ai..Aj (1-based indexing)

    Base:
      dp[i, i] = 0 (one matrix needs no multiplication)

    Transition:
      For i < j, split at k (i <= k < j):
        cost = dp[i,k] + dp[k+1,j] + p[i-1]*p[k]*p[j]

      Why p[i-1]*p[k]*p[j]?
        - (Ai..Ak) produces matrix of size p[i-1] x p[k]
        - (A(k+1)..Aj) produces matrix of size p[k] x p[j]
        - Multiplying them costs p[i-1]*p[k]*p[j]

    Fill order:
      Increasing chain length (len = 2..n) ensures subproblems are ready.

    Overflow note:
      We compute the candidate cost using long because multiplication
      can exceed int range; then we cast back to int after minimizing.
*/
static int MatrixChainOrder(int[] p) {
    int n = p.Length - 1;      // number of matrices
    int INF = int.MaxValue;    // initial "worst" value for minimization

    // dp dimensions: (n+1) x (n+1) so we can use 1..n indices.
    var dp = new int[n + 1, n + 1];

    // len = number of matrices in the subchain (Ai..Aj)
    for (int len = 2; len <= n; len++) {

        // i = start matrix index
        for (int i = 1; i <= n - len + 1; i++) {
            int j = i + len - 1;  // j = end matrix index
            dp[i, j] = INF;

            // try every split point k
            for (int k = i; k < j; k++) {

                // Use long to avoid overflow for the multiplication term.
                long cost =
                    (long)dp[i, k] + dp[k + 1, j] +
                    (long)p[i - 1] * p[k] * p[j];

                // Relax / minimize dp[i,j]
                if (cost < dp[i, j]) {
                    dp[i, j] = (int)cost;
                }
            }
        }
    }

    // Final answer: optimal cost for multiplying A1..An.
    return dp[1, n];
}

// ===========================
// Tests (top-level)
// ===========================

Console.WriteLine("=== Test 1: MaxRootToLeaf ===\n");

/*
    Build tree:

          10
         /  \
        5    20
       / \     \
      3   7     25

    Root-to-leaf sums:
      10+5+3   = 18
      10+5+7   = 22
      10+20+25 = 55  <-- maximum
*/
TreeNode root = new(10)
{
    Left = new TreeNode(5)
    {
        Left = new TreeNode(3),
        Right = new TreeNode(7)
    },
    Right = new TreeNode(20)
    {
        Right = new TreeNode(25)
    }
};

int bestPath = MaxRootToLeaf(root);
Console.WriteLine($"MaxRootToLeaf(tree) = {bestPath} (expected 55)");

/*
    Single node tree:
      - Root is also a leaf, so answer is just its value.
*/
TreeNode single = new(42);
int singleBest = MaxRootToLeaf(single);
Console.WriteLine($"MaxRootToLeaf(single=42) = {singleBest} (expected 42)");

/*
    Null tree:
      - No valid root-to-leaf path, so we return int.MinValue sentinel.
*/
int nullBest = MaxRootToLeaf(null);
Console.WriteLine($"MaxRootToLeaf(null) = {nullBest} (expected {int.MinValue})");

Console.WriteLine();
Console.WriteLine("=== Test 2: TspBitmask ===\n");

/*
    4-node distance matrix (symmetric example).

    We want min-cost path:
      start at 0
      visit all nodes once
      end anywhere

    One optimal path:
      0 -> 1 -> 3 -> 2
      cost = 10 + 25 + 30 = 65
*/
int[,] dist = new int[4, 4]
{
    { 0, 10, 15, 20 },
    { 10, 0, 35, 25 },
    { 15, 35, 0, 30 },
    { 20, 25, 30, 0 }
};

int tspCost = TspBitmask(4, dist);
Console.WriteLine($"TspBitmask(4 nodes) = {tspCost} (expected 65)");

Console.WriteLine();
Console.WriteLine("=== Test 3: MatrixChainOrder ===\n");

/*
    CLRS matrix chain example:
      p = [30, 35, 15, 5, 10, 20, 25]
      n = 6 matrices
      optimal cost = 15125
*/
int[] p1 = { 30, 35, 15, 5, 10, 20, 25 };
int cost1 = MatrixChainOrder(p1);
Console.WriteLine($"MatrixChainOrder(CLRS) = {cost1} (expected 15125)");

/*
    Two matrices:
      A: 10x20
      B: 20x30
    Only one multiplication:
      cost = 10*20*30 = 6000
*/
int[] p2 = { 10, 20, 30 };
int cost2 = MatrixChainOrder(p2);
Console.WriteLine($"MatrixChainOrder([10,20,30]) = {cost2} (expected 6000)");

/*
    Single matrix:
      No multiplication needed => 0 cost.
*/
int[] p3 = { 5, 10 };
int cost3 = MatrixChainOrder(p3);
Console.WriteLine($"MatrixChainOrder(single) = {cost3} (expected 0)");

/* ============================================================
   TreeNode definition (simple binary tree node)
   ============================================================ */

/*
    Small helper class for the tree tests above.

    - Primary constructor sets Val.
    - Left/Right are nullable references (TreeNode?) to support missing children.
    - Fields are public for easy test construction.

    NOTE:
      Using fields (rather than properties) keeps the class minimal for examples.
*/
class TreeNode(int val)
{
    public int Val = val;
    public TreeNode? Left;
    public TreeNode? Right;
}

File: code_samples/section9/lesson/section9.py
"""Lesson 9 - Advanced Dynamic Programming"""

from __future__ import annotations
from dataclasses import dataclass
from typing import Optional

@dataclass
class TreeNode:
    """Model for a binary tree node.

    This is a small, immutable-by-convention data container representing a node
    in a binary tree.

    Attributes:
        val: Integer value stored at this node.
        left: Reference to the left child node, or None if absent.
        right: Reference to the right child node, or None if absent.
    """
    val: int
    left: Optional["TreeNode"] = None
    right: Optional["TreeNode"] = None

def max_root_to_leaf(root: Optional[TreeNode]) -> int:
    """Return the maximum sum of a root-to-leaf path in a binary tree.

    A "root-to-leaf" path:
      - must start at the provided `root` argument
      - must end at some leaf node (a node with no children)

    This is *not* the "maximum path sum anywhere in the tree" problem; the path
    is constrained to begin at the root.

    Approach:
      - Recursively compute the best root-to-leaf path sum for left and right
        subtrees, then take the maximum and add the current node's value.

    Base cases:
      - If root is None: return -infinity to represent "no valid path".
        Using -infinity ensures missing children never win a max() comparison.
      - If root is a leaf: return root.val.

    Args:
        root: The root TreeNode of the (sub)tree, or None.

    Returns:
        The maximum sum along any root-to-leaf path.
        If root is None, returns float("-inf") as a sentinel for "no path".
    """
    # Empty subtree => no path exists. Use -inf so it won't be chosen by max().
    if root is None:
        return float("-inf")

    # Leaf node => the best path ending here is just this node's value.
    if root.left is None and root.right is None:
        return root.val

    # Recursively compute the best sums from each child down to some leaf.
    left_best = max_root_to_leaf(root.left)
    right_best = max_root_to_leaf(root.right)

    # Choose the better child path and include the current node's value.
    return root.val + max(left_best, right_best)

def tsp_bitmask(n: int, dist: list[list[int]]) -> int:
    """Return the minimum cost of a traveling salesman problem.

    This function solves a common "TSP-like" dynamic programming variant:
      - You start at node 0
      - You must visit every node exactly once
      - You may end at any node (this implementation does *not* add a cost to
        return back to node 0)

    DP definition:
        dp[mask][j] = minimum cost to start at 0, visit exactly the nodes whose
                      bits are set in `mask`, and end at node j.

    Initialization:
        dp[1<<0][0] = 0
        (Only node 0 visited, currently at 0, cost 0.)

    Transition:
        From state (mask, j), attempt to go to an unvisited node k:
            next_mask = mask | (1<<k)
            dp[next_mask][k] = min(dp[next_mask][k], dp[mask][j] + dist[j][k])

    Answer:
        min_j dp[full_mask][j], where full_mask has all n bits set.

    Complexity:
        Time:  O(2^n * n^2)
        Space: O(2^n * n)

    Notes:
        - INF is set to a very large value (10**15) to represent unreachable states.
        - The distances in `dist` are assumed to be non-negative and finite.

    Args:
        n: Number of nodes labeled 0..n-1.
        dist: Square matrix dist[i][j] giving travel cost from i to j.

    Returns:
        Minimum cost to start at 0, visit all nodes once, and end anywhere.
    """
    # Total number of subsets of nodes (bitmasks) is 2^n.
    max_mask = 1 << n

    # Sentinel for "unreachable" DP states. Must exceed any realistic tour cost.
    INF = 10**15

    # dp[mask][j] initialized to INF for all states.
    dp = [[INF] * n for _ in range(max_mask)]

    # Starting state: only node 0 visited, currently at node 0, cost 0.
    dp[1 << 0][0] = 0

    # Iterate over all subsets of visited nodes.
    for mask in range(max_mask):
        # Try each possible current endpoint j.
        for j in range(n):
            # If j is not included in mask, then "end at j" is not a valid state.
            if not mask & (1 << j):
                continue

            curr = dp[mask][j]
            # Skip unreachable states.
            if curr == INF:
                continue

            # Try to visit an unvisited node k next.
            for k in range(n):
                # If k is already visited, we cannot visit it again.
                if mask & (1 << k):
                    continue

                next_mask = mask | (1 << k)
                cand = curr + dist[j][k]

                # Relaxation: keep the minimal cost for reaching (next_mask, k).
                if cand < dp[next_mask][k]:
                    dp[next_mask][k] = cand

    # Full mask = all nodes visited.
    full_mask = max_mask - 1

    # End anywhere: take the minimum cost among all endpoints.
    return min(dp[full_mask])

def matrix_chain_order(p: list[int]) -> int:
    """Return the minimum cost of matrix multiplication.

    Matrix Chain Multiplication (MCM) problem:
        Given a chain of matrices A1..An, choose parenthesization to minimize
        scalar multiplications.

    The dimension array `p` encodes matrices:
        A_i has dimensions p[i-1] x p[i] for i = 1..n
    where n = len(p) - 1.

    DP definition (classic CLRS formulation using 1-based matrix indices):
        dp[i][j] = minimum scalar multiplication cost to compute A_i..A_j

    Base:
        dp[i][i] = 0 (a single matrix requires no multiplications)

    Transition:
        dp[i][j] = min over k in [i..j-1] of:
            dp[i][k] + dp[k+1][j] + p[i-1] * p[k] * p[j]

    Complexity:
        Time:  O(n^3)
        Space: O(n^2)

    Notes:
        - INF is a large sentinel to represent "not yet minimized" entries.
        - dp is allocated as (n+1)x(n+1) so indices 1..n are convenient.

    Args:
        p: List of dimensions of length n+1.

    Returns:
        Minimum scalar multiplication cost to compute the full product A1..An.
    """
    # Number of matrices in the chain.
    n = len(p) - 1

    # Sentinel infinity for minimization.
    INF = 10**15

    # dp table: dp[i][j] will be filled for 1 <= i <= j <= n.
    # Defaults to 0; we'll assign INF for i < j when computing.
    dp = [[0] * (n + 1) for _ in range(n + 1)]

    # length is the number of matrices in the subchain.
    # Start at 2 because length=1 subproblems (single matrix) have cost 0.
    for length in range(2, n + 1):
        # i is the starting matrix index for this subproblem.
        for i in range(1, n - length + 2):
            # j is the ending matrix index.
            j = i + length - 1

            # Initialize as INF before taking the min over split points.
            dp[i][j] = INF

            # Try splitting between k and k+1.
            for k in range(i, j):
                # Cost = cost(left subchain) + cost(right subchain) + cost to multiply results.
                # Dimensions:
                #   (A_i..A_k) is p[i-1] x p[k]
                #   (A_{k+1}..A_j) is p[k] x p[j]
                # Multiply cost: p[i-1] * p[k] * p[j]
                cost = dp[i][k] + dp[k + 1][j] + p[i - 1] * p[k] * p[j]

                # Keep the minimal cost among all possible split points.
                if cost < dp[i][j]:
                    dp[i][j] = cost

    # The full chain corresponds to dp[1][n].
    return dp[1][n]

# ===========================
# Tests for Lesson 9
# ===========================

if __name__ == "__main__":
    # Basic smoke tests demonstrating expected outputs for each function.
    print("=== Test 1: max_root_to_leaf ===\n")

    # Build tree:
    #       10
    #      /  \
    #     5    20
    #    / \     \
    #   3   7     25
    #
    # Best root-to-leaf sum = 10 + 20 + 25 = 55
    #
    # Note: Because this is root-to-leaf, paths must start at the root (10).
    root_node = TreeNode(
        10,
        left=TreeNode(
            5,
            left=TreeNode(3),
            right=TreeNode(7),
        ),
        right=TreeNode(
            20,
            right=TreeNode(25),
        ),
    )

    # Expect 55 from the best path 10 -> 20 -> 25.
    best_path = max_root_to_leaf(root_node)
    print(f"max_root_to_leaf(tree) = {best_path} (expected 55)")

    # Single-node tree: that node is also a leaf, so the best sum is its value.
    single = TreeNode(42)
    single_best = max_root_to_leaf(single)
    print(f"max_root_to_leaf(single=42) = {single_best} (expected 42)")

    # Null root: function returns -inf sentinel by design.
    null_best = max_root_to_leaf(None)
    print(f"max_root_to_leaf(None) = {null_best} (expected {float('-inf')})")

    print()
    print("=== Test 2: tsp_bitmask ===\n")

    # 4-node example distances.
    distances = [
        [0, 10, 15, 20],
        [10, 0, 35, 25],
        [15, 35, 0, 30],
        [20, 25, 30, 0],
    ]

    # Expected best cost is 65 for the "start at 0, visit all once, end anywhere" variant.
    tsp_cost = tsp_bitmask(4, distances)
    print(f"tsp_bitmask(4 nodes) = {tsp_cost} (expected 65)")

    print()
    print("=== Test 3: matrix_chain_order ===\n")

    # CLRS example:
    # p = [30,35,15,5,10,20,25] (6 matrices) => expected optimal cost 15125
    p1 = [30, 35, 15, 5, 10, 20, 25]  # 6 matrices
    cost1 = matrix_chain_order(p1)
    print(f"matrix_chain_order(CLRS) = {cost1} (expected 15125)")

    # Simple 2-matrix example: 10x20, 20x30 -> cost = 10*20*30 = 6000
    p2 = [10, 20, 30]
    cost2 = matrix_chain_order(p2)
    print(f"matrix_chain_order([10,20,30]) = {cost2} (expected 6000)")

    # Single matrix: no multiplication needed => cost 0
    p3 = [5, 10]
    cost3 = matrix_chain_order(p3)
    print(f"matrix_chain_order(single) = {cost3} (expected 0)")
